Rewriting the GitHub Event Sample README

n3wscott opened the pull request:
Based on feedback from https://github.com/knative/eventing/pull/276/files#diff-f9667ca3c88cf4316ebe5aab2d712207 I have updated the README to be more in-line with what the current docs repo does.

*Note*: I have also staged the commands for the move to `knative/docs/eventing/samples/github-events`


@evankanderson : I need guidance on how much of the "What just happened" second needs to be pulled from the k8s event or if we can point to that?

Related to https://github.com/knative/eventing/issues/333

Dependent on https://github.com/knative/eventing/pull/340

n3wscott commented:
/assign @evankanderson

n3wscott commented:
Grant for LGTM:
/assign @grantr

Evan for Approval:
/assign @evankanderson

n3wscott commented:
/test pull-knative-eventing-integration-tests

grantr commented:
/cc @grantr

(cc requests my review, so I can track this in a different email label)

evankanderson commented on sample/github/README.md:
````diff
@@ -6,156 +6,130 @@ deployed to receive the webhook's event deliveries and forward them into a
+    legit             True      <none>
+    ```
+
+1.  Create a [personal access token](https://github.com/settings/tokens) to 
````
This needs a little more detail. At a minimum:

1. What scopes should be selected?
2. Does the name matter?
3. Mention that you'll need to copy the string once, and if you forget to/lose it, you'll need to generate a new one.

evankanderson replied:
A screenshot would be nice, feel free to leave as a TODO.

n3wscott replied:
Added a screenshot and a paragraph with the requested info.

evankanderson commented on sample/github/README.md:
```diff
@@ -6,156 +6,130 @@ deployed to receive the webhook's event deliveries and forward them into a
+1.  Create a [personal access token](https://github.com/settings/tokens) to 
+    GitHub repo that the GitHub source can use to register webhooks with the 
+    GitHub API. Also decide on a token that your code will use to authenticate
+    the incoming webhooks from GitHub (*accessToken*). Update 
```
What are the properties that this token should have? Just a unique string?

evankanderson replied:
Maybe there's an easy way to generate a unique 12-character alphanumeric string (`pwgen` or `head -c 8 /dev/urandom | base64encode` might work)?

evankanderson commented on sample/github/README.md:
````diff
@@ -6,156 +6,130 @@ deployed to receive the webhook's event deliveries and forward them into a
+    legit             True      <none>
+    ```
+
+1.  Create a [personal access token](https://github.com/settings/tokens) to 
````
I'd also call out that we use the same token for both the flow (to set up the webhook) and the function (to read and update the PR title).

I'd also move this step up before step 2 (`kubectl apply -f function.yaml`), so that the secrets are in place and the service has everything it needs before step 3 (wait for service to become ready).

evankanderson commented on sample/github/README.md:
```diff
@@ -6,156 +6,130 @@ deployed to receive the webhook's event deliveries and forward them into a
+
+## Understanding what happened
+
+...TODO...
```
I'd make this more formal:

**TODO:** explain the resources and communication channels, as well as where the secret is used.

evankanderson replied:
In particular include a note to look at `https://github.com/<owner>/<repo>/settings/hooks` to see the webhook registered and then deleted.

evankanderson commented on sample/github/README.md:
````diff
@@ -6,156 +6,130 @@ deployed to receive the webhook's event deliveries and forward them into a
-```
-
-And you can check the repo and see the webhook has been removed
+To clean up the function, `Flow`, auth, and secret:
````
Also mention deleting the github Personal Access Token from the github UI.

greghaynes commented on sample/github/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+FROM golang AS builder
+
+WORKDIR /go/src/github.com/knative/docs/
+ADD . /go/src/github.com/knative/docs/
```
Dockerfile pathing is relative to the dockerfile itself which causes two issues here:

- this github-events directory's contents gets placed under /go/src/github.com/knative/docs/ which results in an unable to find eventing/samples/github-events error when compiling. This can be fixed by changing the above command to `ADD . /go/src/github.com/knative/docs/eventing/samples/github-events`
- Because we are only copying the github-events dir we lack the deps for this example (which are in docs/vendor). ISTM we can either add them with go get or add dep to the docker image along with a Gopkg config and then run dep ensure.

n3wscott replied:
Thanks for the tip! I will be sure to fix this up.

greghaynes commented:
The current function.yaml is written such that it requires use of Ko / doesnt line up with the README as the image: value will be incorrect. Not sure what the best fix is here - we could ask a user to edit this value to match their username/github-events?

n3wscott commented on sample/github/README.md:
````diff
@@ -6,156 +6,130 @@ deployed to receive the webhook's event deliveries and forward them into a
-```
-
-Then create the secret so that you can see changes:
+The `eventing/samples/github-events/auth.yaml` file provisions a service account, creates a role
````
The docs are written to run from the root of the docs directory so I have already changed things to make that assumption: `eventing/samples/github-events`

n3wscott commented on sample/github/README.md:
````diff
@@ -6,156 +6,130 @@ deployed to receive the webhook's event deliveries and forward them into a
 
 ```shell
- ko apply -f sample/github/githubsecret.yaml
+kubectl apply -f eventing/samples/github-events/auth.yaml
````
I was following the docs repo pattern. (as soon as this lands I am deleting the files from this repo.

n3wscott commented on sample/github/README.md:
```diff
@@ -6,156 +6,130 @@ deployed to receive the webhook's event deliveries and forward them into a
+
+1.  Use Docker to build the sample code into a container. To build and push with
+    Docker Hub, run these commands replacing `{username}` with your Docker Hub
+    username. Run the following from the _root_ of the `knative/docs` repo:
```
again, I am preparing for the move, figured I would edit it once

n3wscott commented:
Updated, PTAL @evankanderson

n3wscott commented:
Follow-up move to docs: https://github.com/knative/docs/pull/339

n3wscott commented:
Follow-up to delete the github samples from eventing repo: https://github.com/knative/eventing/pull/383

grantr commented on sample/github/README.md:
````diff
@@ -6,156 +6,156 @@ deployed to receive the webhook's event deliveries and forward them into a
-# This will show the Revision that was created by the Configuration:
-kubectl get revisions -o yaml
-```
+- A Kubernetes cluster with Knative installed. Follow the
````
This should probably be `with Knative serving installed`.

grantr commented on sample/github/README.md:
```diff
@@ -6,156 +6,156 @@ deployed to receive the webhook's event deliveries and forward them into a
+  if you need to create one.
+- [Docker](https://www.docker.com/) installed and running on your local machine,
+  and a Docker Hub account configured (you'll use it for a container registry).
+- The core Knative eventing tools installed. You can install them with:
```
I don't think I'd call these "tools". Maybe `The Knative eventing core installed.`? Alternately, amend the above to say `A Kubernetes cluster with Knative serving and eventing installed`.

grantr commented on sample/github/README.md:
````diff
@@ -6,156 +6,156 @@ deployed to receive the webhook's event deliveries and forward them into a
+  ```shell
+  kubectl apply -f https://storage.googleapis.com/knative-releases/eventing/latest/release.yaml
+  ```
+- A custom domain is setup to allow for GitHub to be able to call into the cluster:
````
The grammar here doesn't line up with the other bullets. I'd rephrase this to `A domain name that allows GitHub to call into the cluster`.

n3wscott commented:
I had to use the code from https://github.com/knative/eventing/pull/339

Closes https://github.com/knative/eventing/pull/339

Readme works. Source had two minor issues:

- log flag was missing because glog is not used. removed.
- istio sidecar injection was preventing the feedlet job from starting. fixed from patch

grantr commented on sample/github/README.md:
````diff
@@ -6,156 +6,156 @@ deployed to receive the webhook's event deliveries and forward them into a
-ko apply -f sample/github/flow.yaml
-```   
+Because the `github` EventSource needs to create a
+`Service.serving.knative.dev`, you'll need to provision a special
````
I'd recommend using the non-fully scoped name here: `EventSource needs to create a Knative Service`

grantr commented on sample/github/README.md:
```diff
@@ -6,156 +6,156 @@ deployed to receive the webhook's event deliveries and forward them into a
+    The token can be named anything you find convenient. This sample requires
+    full `repo` control to be able update the title of the _Pull Request_.
+    The Source requires `admin:repo_hook`, this allows it to create webhooks
+    into repos that your account is allowed to do so. Copy and save this token,
```
Replace the comma with a semicolon: `Copy and save this token; GitHub will...`

n3wscott commented:
Updated both https://github.com/knative/docs/pull/339 and here.

n3wscott commented:
/test pull-knative-eventing-integration-tests

n3wscott commented:
/test pull-knative-eventing-integration-tests

grantr commented:
/assign @evankanderson

n3wscott commented:
Closed in favor of https://github.com/knative/docs/pull/339
